pycbc/events/coinc.py

```diff
@@ -531,126 +532,59 @@ def __init__(self, num_rings, max_length, dtype=numpy.float32):
-        self.pad_count = size
-        self.buffer = numpy.resize(self.buffer, (self.num_rings, size))
-        self.buffer_expire = numpy.resize(self.buffer_expire, (self.num_rings, size))
-
-    @property
-    def start_time(self):
-        return self.buffer[0][self.start[0]]['end_time']
+        self.max_time = max_time
+        self.buffer = []
+        self.buffer_expire = []
+        for i in range(num_rings):
+            self.buffer.append([])
+            self.buffer_expire.append([])
+        self.time = 0
 
     @property
-    def end_time(self):
-        return self.buffer[0][self.index[0]]['end_time']
-
-    def ring_sizes(self):
-        count = self.index - self.start
-        count[self.index < self.start] += self.pad_count
-        return count
+    def filled_time(self):
+        return min(self.time, self.max_time)
 
     def num_elements(self):
-        total = self.ring_sizes().sum()
-        return total
+        return sum([len(a) for a in self.buffer])
 
     @property
     def nbytes(self):
-        return self.buffer.nbytes
+        return sum([a.nbytes for a in self.buffer])
 
     def discard_last(self, indices):
         """Discard the triggers added in the latest update"""
-        index = self.index[indices]
-        index -= 1
-        index[index < 0] = self.pad_count - 1
-        self.index[indices] = index
+        for i in indices:
+            self.buffer_expire[i] = self.buffer_expire[i][:-1]
+            self.buffer[i] = self.buffer[i][:-1]
 
     def advance_time(self):
         """Advance the internal time inrement by 1, expiring any triggers that
         are now too old.
         """
-        if self.size < self.max_length:
-            self.size += 1
-        self.expire += 1
+        self.time += 1
 
-        idx = self.buffer_expire[self.ladder, self.start] < self.expire - self.max_length
-        self.start[numpy.logical_and(idx, self.start != self.index)] += 1
-        self.start[self.start >= self.pad_count] = 0
+        expired = self.time - self.max_time
+        for j, exp in enumerate(self.buffer_expire):
+            if (len(exp) > 0) and (exp[0] < expired):
+                self.buffer_expire[j] = exp[1:].copy()
+                self.buffer[j] = self.buffer[j][1:].copy()
 
     def add(self, indices, values):
         """Add triggers in 'values' to the buffers indicated by the indices
         """
-        if self.ring_sizes().max() + 2 > self.pad_count * .9:
-            self.increase_buffer_size(int(self.pad_count * 1.5 + 5))
-
-        index = self.index[indices]
-
-        self.buffer[indices, index] = values
-        self.buffer_expire[indices, index] = self.expire
-
-        index += 1
-        index[index >= self.pad_count] = 0
-        self.index[indices] = index
+        for i, v in zip(indices, values):
+            self.buffer[i] = numpy.concatenate([self.buffer[i], v])
+            self.buffer_expire[i] = numpy.concatenate([self.buffer[i], [self.time]])
```

Comment:
I just spotted this trying to address another error. Should this line not read
```
self.buffer_expire[i] = numpy.concatenate([self.buffer_expire[i], [self.time]])
```
?

Reply:
Uh.... yes, it should!

Reply:
Ok, I can commit that fix.